internal/cmd/envoy/shutdown_manager.go

```diff
@@ -0,0 +1,209 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package envoy
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/envoyproxy/gateway/api/v1alpha1"
+	"github.com/envoyproxy/gateway/internal/logging"
+	"github.com/envoyproxy/gateway/internal/xds/bootstrap"
+)
+
+var (
+	logger = logging.DefaultLogger(v1alpha1.LogLevelInfo).WithName("shutdown-manager")
+)
+
+const (
+	// ShutdownReadyPort is the port Envoy shutdown manager will listen on.
+	ShutdownReadyPort = 19002
+	// ShutdownReadyFile is the file used to indicate shutdown readiness.
+	ShutdownReadyFile = "/tmp/shutdown-ready"
+)
+
+// ShutdownManager serves shutdown manager process for Envoy proxies.
+func ShutdownManager(readyTimeout time.Duration) error {
+	// Setup HTTP handler
+	handler := http.NewServeMux()
+	handler.HandleFunc("/shutdown/ready", func(w http.ResponseWriter, r *http.Request) {
+		shutdownReadyHandler(w, r, ShutdownReadyFile)
```

Comment:
Can this timeout before the handler actually detects shutdown readdiness due to the server timeouts and/or kubelet timeouts? If so, what would be the behavior?

Reply:
It has a [configurable timeout](https://github.com/envoyproxy/gateway/pull/2633/files#diff-7a44a61d8a97c19e682d34666a5b0616719ef5a0ab3fa0b857a4863d8294e5f0R67-R68
) (defaulting to 10s longer than the drain timeout) as a safeguard, but it won't timeout prior to this per documented and observed behavior.

If this did return before the `exec` prestop completed the drain, envoy would be sent a SIGTERM and exit after which you may observe errors such as this one logged until the drain timeout elapses and the shutdown command exits:

```
shutdown-manager 2024-02-22T20:48:54.259Z    ERROR    shutdown-manager    envoy/shutdown_manager.go:145    error getting total connections    {"error": "Get \"http://127.0.0.1:19000//stats?filter=^server\\\\.total_connections$&format=json\": dial tcp 127.0.0.1:19000: connect: connection refused"}
```

This should never occur with properly configured (staggered) timeouts, so I would hesitate to code around it for an initial implementation preferring to let the errors be logged to get eyes on the potential problem.